src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +148,331 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
+#endif // ASSERT
+}
+
+// Stack used when performing DFS on a Template Assertion Predicate Bool. The DFS traversal visits non-CFG inputs of a
+// node in increasing node index order (i.e. first visiting the input node at index 1). Each time a new node is visited,
+// it is inserted on top of the stack. Each node n in the stack maintains a node input index i, denoted as [n, i]:
+//
+// i = 0, [n, 0]:
+//     n is currently being visited for the first time in the DFS traversal and was newly added to the stack
+//
+// i > 0, [n, i]:
+//     Let node s be the next node being visited after node n in the DFS traversal. The following holds:
+//         n->in(i) = s
+class DFSNodeStack : public StackObj {
+  Node_Stack _stack;
+  static const uint _no_inputs_visited_yet = 0;
+
+ public:
+  explicit DFSNodeStack(BoolNode* template_bool)
+      : _stack(2) {
+        _stack.push(template_bool, _no_inputs_visited_yet);
+  }
+
+  // Push the next unvisited input of the current node on top of the stack and return true. The visiting order of node
+  // inputs is in increasing node input index order. If there are no unvisited inputs left, do nothing and return false.
+  bool push_next_unvisited_input() {
+    Node* node_on_top = top();
+    increment_top_node_input_index();
+    const uint next_unvisited_input = _stack.index();
+    for (uint index = next_unvisited_input; index < node_on_top->req(); index++) {
+      Node* input = node_on_top->in(index);
+      if (TemplateAssertionPredicateBool::could_be_part(input)) {
+        // We only care about nodes that could possibly be part of a Template Assertion Predicate Bool.
+        _stack.set_index(index);
+        _stack.push(input, _no_inputs_visited_yet);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  Node* top() const {
+    return _stack.node();
+  }
+
+  uint node_index_to_previously_visited_parent() const {
+    return _stack.index();
+  }
+
+  bool is_not_empty() const {
+    return _stack.size() > 0;
+  }
+
+  Node* pop() {
+    Node* popped_node = top();
+    _stack.pop();
+    return popped_node;
+  }
+
+  void increment_top_node_input_index() {
+    _stack.set_index(_stack.index() + 1);
+  }
+
+  void replace_top_with(Node* node) {
+    _stack.set_node(node);
+  }
+};
+
+// Interface to transform OpaqueLoop* nodes of a Template Assertion Predicate Bool. The transformations must return a
+// new or different existing node.
+class TransformOpaqueLoopNodes : public StackObj {
+ public:
+  virtual Node* transform_opaque_init(OpaqueLoopInitNode* opaque_init) = 0;
+  virtual Node* transform_opaque_stride(OpaqueLoopStrideNode* opaque_stride) = 0;
+};
+
+// Class to clone a Template Assertion Predicate Bool. The BoolNode and all the nodes up to but excluding the OpaqueLoop*
+// nodes are cloned. The OpaqueLoop* nodes are transformed by the provided strategy (e.g. cloned or replaced).
+class CloneTemplateAssertionPredicateBool : public StackObj {
```

Comment:
```suggestion
class CloneTemplateAssertionPredicateExpression : public StackObj {
```

Reply:
So we need to do work whenever we walk back down a input->output edge. We can traverse this edge, by taking the post-order traversal, and then waling from the post-order visited node to its output node.

```
    Node* current;
    while (_stack.is_not_empty()) {
      current = _stack.top();
      if (current->is_Opaque1()) {
        Node* transformed_node = transform_opaque_loop_node(current, transform_opaque_loop_nodes);
        _stack.replace_top(transformed_node);
        traverse_edge_back_to_output_and_pop();
      } else if (!_stack.push_next_unvisited_input()) {
        traverse_edge_back_to_output_and_pop();
      } // else: we just pushed an new input, go and visit it first
    }
    
traverse_edge_back_to_output_and_pop:
  Node* maybe_cloned_input = _stack.pop();
  if (_stack.is_empty()) {
    return; // we just visited the root node, and are now done. Maybe verify we have the root here?  
  }
  Node* output = _stack.top();
  int i = _stack.top_index();
  if (!is_clone(output)) {
    output = output->clone();
    _stack.replace_top(output);
  }
  output->set_req(i, maybe_cloned_input);
```

Reply:
An even easier algorithm would be to have an additional `old_new` data-structure that knows the clones for the old nodes.

Then you simply post-order traverse, and whenever you post-visit a node (current, all its inputs were already traversed, and if need be cloned), you check all the inputs, and see if any of them were cloned. If so, clone current also, and add it to `old_new`. That would mean that you do not have to hack the DFS iterator, and could use it in read-only mode. And it would be a much simpler algorithm. But the whole thing comes as the cost of this extra `old_new` data-structure.